### ISSUE TITLE
Referencing Component inside getServerSideProps breaks build (Next.js 12)

### ISSUE THREAD

**Issue by santialbo** [vercel/next.js#30611]
### What version of Next.js are you using?

12.0.3

### What version of Node.js are you using?

16.6.1

### What browser are you using?

Chrome, Firefox

### What operating system are you using?

macOS

### How are you deploying your application?

Doesn't matter

### Describe the Bug

After upgrading my project to Next.js 12 the generated builds are not correct. I've identified the following issue:

If a component is referenced inside `getServerSideProps` as shown below the generated build is incorrect and gives client side errors. **This issue didn't happen on Next 11**.

```tsx
import { ComponentFoo } from "../components/ComponentFoo";

export default function Home() {
  return (
    <div>
      <ComponentFoo />
    </div>
  );
}

export function getServerSideProps() {
  return {
    props: { foo: ComponentFoo.fragments.foo },
  };
}
```

I've created a minimal reproduction [here](https://github.com/santialbo/test-next-build) which is deployed on vercel [here](https://test-next-build-ozkc28jk0-santialbo.vercel.app/).

Basically the error is that the generated build for the client is not including the component definition. See comparison
.
Referencing the component from inside getServerSideProps:
<img width="695" alt="Screenshot 2021-10-29 at 11 14 25" src="https://user-images.githubusercontent.com/1557563/139412355-7fbcdb19-1a65-49de-b679-e4d69d15400a.png">

**Not** referencing the component from inside getServerSideProps:
<img width="695" alt="Screenshot 2021-10-29 at 11 19 10" src="https://user-images.githubusercontent.com/1557563/139412283-dc8438a8-6673-4560-bcf7-923416b8f541.png">

This is generating a `ComponentFoo is not defined` error when rehydrating on the client.

*Reason why I'm referencing the components*:
I attach GraphQL fragments to the components specifying the data they need to render. When calling getServerSideProps I need to reference those fragments in order to fetch the necessary data.

### Expected Behavior

I would it expect to work. The same way it worked on Next 11.

### To Reproduce

Minimal reproduction: https://github.com/santialbo/test-next-build
Vercel deployment: https://test-next-build.vercel.app/

**Comment by benknight** [comment]
Thank you for reporting this issue. I am experiencing exactly the same problem after updating to version 12.

I am using the same pattern as OP where I'm defining GraphQL fragments as fields on my exported components.

**Comment by benknight** [comment]
Update - 2 months later and I still can't upgrade to the latest version of Next because of this bug. Hoping for a fix soon.

**Comment by santialbo** [comment]
> Update - 2 months since this report and I still can't upgrade to the latest version of Next because of this bug. Hoping for a fix soon!
> 
> I considered refactoring my code to not import fragments this way, but I'm pretty sure this is a pattern used by a lot of React/GraphQL projects and therefore Next should accommodate it. I think it's common because it's even recommended here in the Apollo docs: https://www.apollographql.com/docs/react/data/fragments/#colocating-fragments

I upgraded to v12 by creating a dummy .babelrc file. This makes next opt-out of using SWC. https://nextjs.org/docs/messages/swc-disabled
It's not ideal but at least you can use the rest of v12 features.

**Comment by benknight** [comment]
I ended up just refactoring my code and pulling out all my fragments into a
separate file. It's unfortunate because colocation is nice but ultimately
not really a big deal.

**Comment by santialbo** [comment]
I really like collocation of fragments, I wouldn't like to give up on them. This bug is just a symptom of an underlying issue with the client side code elimination which I hope can be fixed soon.

**Comment by balazsorban44** [comment]
Hi, this has been fixed and is working now, just tested with `12.0.8`. Please upgrade.

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
